src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,157 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.NoSuchElementException;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (size == 0) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(value);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkPositionIndex(index);
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index - 1);
+            newNode.next = nodeByIndex.next;
+            nodeByIndex.next.prev = newNode;
+            nodeByIndex.next = newNode;
+            newNode.prev = nodeByIndex;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.size() == 0) {
+            throw new NullPointerException("This list is empty" + list);
+        }
```

Comment:
It would be okay if you do not add anything if the list is empty. NPE must be thrown only when something is null